cdap-security/src/test/java/co/cask/cdap/security/authorization/InMemoryAuthorizer.java

```diff
@@ -168,24 +172,33 @@ public void removeRoleFromPrincipal(Role role, Principal principal) throws RoleN
         privileges.addAll(getPrivileges(role));
       }
     }
-    return privileges;
+    return Collections.unmodifiableSet(privileges);
   }
 
   private Set<Privilege> getPrivileges(Principal principal) {
-    Set<Privilege> privileges = new HashSet<>();
-    Set<Map.Entry<EntityId, Set<Action>>> entries = table.column(principal).entrySet();
-    for (Map.Entry<EntityId, Set<Action>> entry : entries) {
-      for (Action action : entry.getValue()) {
-        privileges.add(new Privilege(entry.getKey(), action));
+    Set<Privilege> result = new HashSet<>();
+    for (Map.Entry<EntityId, ConcurrentMap<Principal, Set<Action>>> entry : privileges.entrySet()) {
+      EntityId entityId = entry.getKey();
+      ConcurrentMap<Principal, Set<Action>> value = entry.getValue();
+      if (!value.containsKey(principal)) {
+        continue;
+      }
+      Set<Action> actions = value.get(principal);
+      for (Action action : actions) {
+        result.add(new Privilege(entityId, action));
       }
     }
-    return privileges;
+    return Collections.unmodifiableSet(result);
   }
 
   private Set<Action> getActions(EntityId entity, Principal principal) {
```

Comment:
@chtyim this convenience method helps because it is needed at 4 other places - grant, revoke, list privileges of a principal and list privileges of a role. I have changed it per my previous comment. Could you please check to see if it is better now?